File: merge_intervals.py
# Given a collection of intervals, merge all overlapping intervals.

# example:
# [[2,3],[2,6],[8,10]] => [[1,6],[8,10]]

# challenge: 
# O(n log n) time and O(1) extra space.



"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end

class Solution(object):
    # @param intervals, a list of Interval
    # @return a list of Interval

    """
    It takes me almost much time to debug it on the OJ.
    The main though behind this is first to sort the 
    original intervals by the start, then we are keeping
    scaning in the intervals.
    Notice 
    ss: is to track the start value of the final interval
    value: end value for the final interval
    down :is the index we want to compare with  value
    """

    def merge(self, intervals):
        # write your code here

        if not intervals:
        	return []
        res=[]
        intervals.sort(key=lambda x:x.start)
        return intervals[0]
        ss=0
        down=0
        value=intervals[ss].end
        while down < len(intervals):
        	# append the inteval since we are sure it has no 
        	# overlap with previous value
        	if value < intervals[down].start:
        		res.append([intervals[ss].start,value])
        		ss=down # be care of the update, cost me time to debug
        		value=intervals[down].end
        	# update the value
        	else:
        	    if value <= intervals[down].end:
        		    value=intervals[down].end
        	down+=1
        # jump out the loop, then we need to append last inteval
        res.append([intervals[ss].start,value])
        # i get the wrong return value, although it pass the OJ,
        # the problem want to return the list of intervals which 
        # is list of object.
        return res



# the same idea as me, but get the right output
class Solution2(object):
    def merge(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: List[Interval]
        """
        lenInter = len(intervals)
        if lenInter == 0:
            return intervals

        #Another sorting method: intervals = sorted(intervals, key=lambda inter: inter.start)
        intervals.sort(key=lambda inter: inter.start)

        rsIntervals = []

        curInterval = intervals[0]
        for i in range(1, lenInter):
            if intervals[i].start > curInterval.end:
                rsIntervals.append(curInterval)
                curInterval = intervals[i]
            else:
                curInterval.end = max(curInterval.end, intervals[i].end)

        rsIntervals.append(curInterval)

        return rsIntervals



if __name__=='__main__':
	a,b,c=Interval(1,4),Interval(0,10),Interval(3,5)
	test=[a,b,c]
	A=Solution()
	print A.merge(test)

File: lintcodeInJava/MergeIntervals.java
/*
Given a collection of intervals, merge all overlapping intervals.

For example,
Given [1,3],[2,6],[8,10],[15,18],
return [1,6],[8,10],[15,18].
*/

// copy from others, 

/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */
 
/*
some classifications for this problem:
1: are all intervals given are valid? 
2: how about the case where two intervals has the leftBound = rightBound, like [1,2],[2,3]
3: Can we use the sort function in the Collections module.
*/

public class Solution {
    public List<Interval> merge(List<Interval> intervals) {
           if (intervals == null || intervals.size() <= 1)
              return intervals;
   
          // sort intervals by using self-defined Comparator
          Collections.sort(intervals, new IntervalComparator());
   
          ArrayList<Interval> result = new ArrayList<Interval>();
   
   		//根据已排好序的list，依次遍历每个元素，进行合并
         Interval prev = intervals.get(0);
         for (int i = 1; i < intervals.size(); i++) {
             Interval curr = intervals.get(i);
  
             if (prev.end >= curr.start) {
                 // merged case
                 Interval merged = new Interval(prev.start, Math.max(prev.end, curr.end));
                 prev = merged;
             } else {
                 result.add(prev);
                 prev = curr;
             }
         }
  
         result.add(prev);
  
         return result;
         }
     }
  
  	//此题的重点是重写comparator函数来对原list进行排序，排序的规则是根据左边节点的大小，这边我们默认
     // 所有的interval都是合理的
     class IntervalComparator implements Comparator<Interval> {
         public int compare(Interval i1, Interval i2) {
             return i1.start - i2.start;
         }
}